lib/rules/id-blacklist.js

```diff
@@ -35,88 +134,64 @@ module.exports = {
+         * These identifiers will be allowed, as it is assumed that user has no control over the names of external global variables.
          * @param {ASTNode} node `Identifier` node to check.
-         * @returns {boolean} `true` if the node is a renamed import.
+         * @returns {boolean} `true` if the node is a reference to a global variable.
          */
-        function isRenamedImport(node) {
-            const parent = node.parent;
+        function isReferenceToGlobalVariable(node) {
+            const variable = globalScope.set.get(node.name);
 
-            return (
-                (
-                    parent.type === "ImportSpecifier" &&
-                    parent.imported !== parent.local &&
-                    parent.imported === node
-                ) ||
-                (
-                    parent.type === "ExportSpecifier" &&
-                    parent.parent.source && // re-export
-                    parent.local !== parent.exported &&
-                    parent.local === node
-                )
-            );
+            return variable && variable.defs.length === 0 &&
+                variable.references.some(ref => ref.identifier === node);
         }
 
         /**
-         * Checks whether the given node is a renamed identifier node in an ObjectPattern destructuring.
-         *
-         * Examples:
-         * const { a : b } = foo; // node `a` is renamed node.
-         * @param {ASTNode} node `Identifier` node to check.
-         * @returns {boolean} `true` if the node is a renamed node in an ObjectPattern destructuring.
+         * Determines whether the given node should be checked.
+         * @param {ASTNode} node `Identifier` node.
+         * @returns {boolean} `true` if the node should be checked.
          */
-        function isRenamedInDestructuring(node) {
+        function shouldCheck(node) {
             const parent = node.parent;
 
-            return (
-                (
-                    !parent.computed &&
-                    parent.type === "Property" &&
-                    parent.parent.type === "ObjectPattern" &&
-                    parent.value !== node &&
-                    parent.key === node
-                )
-            );
-        }
-
-        /**
-         * Verifies if we should report an error or not.
-         * @param {ASTNode} node The node to check
-         * @returns {boolean} whether an error should be reported or not
-         */
-        function shouldReport(node) {
-            const parent = node.parent;
+            /*
+             * Member access has special rules for checking property names.
+             * Read access to a property with a blacklisted name is allowed, because it can be on an object that user has no control over.
+             * Write access isn't allowed, because it potentially creates a new property with a blacklisted name.
+             */
+            if (
+                parent.type === "MemberExpression" &&
```

Comment:
maybe, Is it intended to ignore an identifier in assignment without declaration?

I tested it and recognized it pass the blacklist identifier with assignment in destructuring.

This is the case which I tested.
```js
({a: parseInt} = a); // no warning in this pr version.

// but it warns with the declaration.
var paseInt; ({a: parseInt} = a); 
```
And it is warned in v6.8.0 event without declaraion.
* [online demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgaWQtYmxhY2tsaXN0OiBbXCJlcnJvclwiLCBcInBhcnNlSW50XCJdICovXG5cbnBhcnNlSW50ID0gXCJkb29cIjtcblxuKHthOiBwYXJzZUludH0gPSBmb28pOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTEsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnsiYnJvd3NlciI6ZmFsc2UsIm5vZGUiOmZhbHNlfX19)

```js
/*eslint id-blacklist: ["error", "parseInt"] */
({a: parseInt} = foo);
```

Reply:
> maybe, Is this PR intended to ignoring check an identifier in an assignment that has no declaration?

Only if it is a global variable that isn't declared in the same source code. The logic is: if it is declared, then we can assume it's a user-controlled name. Otherwise, it's a built-in/env- specific/third-party global so user can't change the name. 

It could be also indeed a user-controlled name declared elsewhere, but in that case `id-blacklist` will warn in the file where it is declared.

This shouldn't be a warning because `parseInt` is a [global variable](https://github.com/sindresorhus/globals/blob/92bbb85fc797e03465fbf6e69fb881be4f07e23d/globals.json#L92) that isn't declared in this code:

```js
/*eslint id-blacklist: ["error", "parseInt"] */

({a: parseInt} = a);
```

This should be warning (2 warnings, one for each `Identifier` node) because `parseInt` is declared:

```js
/*eslint id-blacklist: ["error", "parseInt"] */

var parseInt; ({a: parseInt} = a);
```

This should be also a warning because foo isn't defined in the configuration as a global variable:

```js
/*eslint id-blacklist: ["error", "foo"] */

({a: foo} = a);
```

Reply:
@mdjermanovic 
> Only if it is a global variable that isn't declared in the same source code. The logic is: if it is declared, then we can assume it's a user-controlled name. Otherwise, it's a built-in/env- specific/third-party global so user can't change the name.

Thanks for the explanation. 👍 it makes sense to me.